hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/RDBBatchOperation.java

```diff
@@ -80,26 +81,33 @@ private static String countSize2String(int count, long size) {
    * To implement {@link #equals(Object)} and {@link #hashCode()}
    * based on the contents of the bytes.
    */
-  static final class Bytes {
-    private final byte[] array;
-    private final CodecBuffer buffer;
+  static final class Bytes implements Closeable {
+    private AbstractSlice<?> slice;
     /** Cache the hash value. */
-    private final int hash;
+    private int hash;
 
     Bytes(CodecBuffer buffer) {
-      this.array = null;
-      this.buffer = Objects.requireNonNull(buffer, "buffer == null");
-      this.hash = buffer.asReadOnlyByteBuffer().hashCode();
+      Objects.requireNonNull(buffer, "buffer == null");
+      if (buffer.isDirect()) {
+        initWithDirectByteBuffer(buffer.asReadOnlyByteBuffer());
+      } else {
+        initWithByteArray(buffer.getArray());
+      }
     }
 
     Bytes(byte[] array) {
-      this.array = array;
-      this.buffer = null;
+      Objects.requireNonNull(array, "array == null");
+      initWithByteArray(array);
+    }
+
+    private void initWithByteArray(byte[] array) {
+      this.slice = new ManagedSlice(array);
       this.hash = ByteBuffer.wrap(array).hashCode();
     }
 
-    ByteBuffer asReadOnlyByteBuffer() {
-      return buffer.asReadOnlyByteBuffer();
+    private void initWithDirectByteBuffer(ByteBuffer byteBuffer) {
+      this.slice = new ManagedDirectSlice(byteBuffer);
+      this.hash = byteBuffer.hashCode();
     }
```

Comment:
Constructors should not call any functions.  Otherwise, the fields cannot be final, which is very useful for understanding the code..
```java
    private final AbstractSlice<?> slice;
    /** Cache the hash value. */
    private final int hash;

    Bytes(CodecBuffer buffer) {
      Objects.requireNonNull(buffer, "buffer == null");
      if (buffer.isDirect()) {
        this.slice = new ManagedDirectSlice(buffer.asReadOnlyByteBuffer());
      } else {
        this.slice = new ManagedSlice(buffer.getArray());
      }
      this.hash = buffer.hashCode();
    }

    Bytes(byte[] array) {
      Objects.requireNonNull(array, "array == null");
      this.slice = new ManagedSlice(array);
      this.hash = ByteBuffer.wrap(array).hashCode();
    }
```

Reply:
done